diff --git a/droidmediaconvert.cpp b/droidmediaconvert.cpp
index e4b0581..f633e7c 100644
--- a/droidmediaconvert.cpp
+++ b/droidmediaconvert.cpp
@@ -40,6 +40,7 @@ public:
 
     ~_DroidMediaConvert() {
         if (m_handle) {
+            closeColorConverterLib();
             dlclose(m_handle);
             m_handle = NULL;
         }
@@ -67,6 +68,8 @@ public:
 
         func(this);
 
+        openColorConverterLib();
+
         return true;
     }
 

