diff --git a/.config-full b/.config-full
index bca8c5eec..5db2716a0 100644
--- a/.config-full
+++ b/.config-full
@@ -548,7 +548,7 @@ CONFIG_FEATURE_FDISK_WRITABLE=y
 # CONFIG_FEATURE_SGI_LABEL is not set
 # CONFIG_FEATURE_SUN_LABEL is not set
 # CONFIG_FEATURE_OSF_LABEL is not set
-# CONFIG_FEATURE_GPT_LABEL is not set
+CONFIG_FEATURE_GPT_LABEL=y
 # CONFIG_FEATURE_FDISK_ADVANCED is not set
 # CONFIG_FINDFS is not set
 CONFIG_FLOCK=y
