diff --git a/init-script b/init-script
index b43d69b..38c81f8 100755
--- a/init-script
+++ b/init-script
@@ -171,6 +171,10 @@ write() {
   echo -n "$2" > "$1"
 }
 
+dmesg_info() {
+    echo "[hybris-boot] $@" > /dev/kmsg
+}
+
 inject_loop() {
     INJ_DIR=/init-ctl
     INJ_STDIN=$INJ_DIR/stdin
@@ -343,12 +347,18 @@ if [ "$DONE_SWITCH" = "no" ]; then
 
     do_mount_devprocsys
 
+    dmesg_info "Hello from the initramfs"
+
     do_hotplug_scan
 
     # Support /dev/block/mmcXXX only in initrd phase
     ln -s . /dev/block
     ln -s /proc/mounts /etc/mtab
 
+    # Handle whole moto_msm8960_jbbl family automatically
+    DATA_PARTITION="$(fdisk -l /dev/mmcblk0 | awk '/userdata/ { print "/dev/mmcblk0p"$1 }')"
+    dmesg_info "userdata=$DATA_PARTITION"
+
     check_kernel_config
 
     bootsplash
@@ -386,7 +396,12 @@ if [ "$DONE_SWITCH" = "no" ]; then
     echo "" > /proc/sys/kernel/hotplug
 
     if [ -f "/target/init-debug" ]; then
-	exec switch_root /target /init-debug &> /target/init-debug-stderrout
+        if [ ! -f "/target/etc/systemd/system/system.mount.d/block-device.conf" ]; then
+            SYSTEM_PARTITION="$(fdisk -l /dev/mmcblk0 | awk '/system/ { print "/dev/mmcblk0p"$1 }')"
+            echo -e "[Mount]\nWhat=$SYSTEM_PARTITION" > /target/etc/systemd/system/system.mount.d/block-device.conf
+            dmesg_info "system=$SYSTEM_PARTITION"
+        fi
+        exec switch_root /target /init-debug &> /target/init-debug-stderrout
     else
 	# Prefer /sbin/preinit over /sbin/init
 	[ -x /target/sbin/preinit ] && INIT=/sbin/preinit || INIT=/sbin/init
@@ -403,6 +418,8 @@ else
 
     do_mount_devprocsys
 
+    dmesg_info "init-debug"
+
     HALT_BOOT="n"
     [ -f /init_enter_debug2 ] && HALT_BOOT="y"
     [ -f /init_disable_telnet ] && DISABLE_TELNET="y"
