diff --git a/init-script b/init-script
index b43d69b..f860d1e 100755
--- a/init-script
+++ b/init-script
@@ -171,6 +171,10 @@ write() {
   echo -n "$2" > "$1"
 }
 
+dmesg_info() {
+    echo "[hybris-boot] $@" > /dev/kmsg
+}
+
 inject_loop() {
     INJ_DIR=/init-ctl
     INJ_STDIN=$INJ_DIR/stdin
@@ -333,6 +337,15 @@ check_kernel_config() {
     fi
 }
 
+adjust_block_device() {
+    conf_path="/target/etc/systemd/system/$1.mount.d/block-device.conf"
+    if [ ! -f "$conf_path" ]; then
+        block_device="$(fdisk -l /dev/mmcblk0 | awk '/'$1'/ { print "/dev/mmcblk0p"$1 }')"
+        echo -e "[Mount]\nWhat=$block_device" > $conf_path
+        dmesg_info "$1=$block_device"
+    fi
+}
+
 # Now either initrd or rootfs sequence
 
 if [ "$DONE_SWITCH" = "no" ]; then
@@ -343,12 +356,18 @@ if [ "$DONE_SWITCH" = "no" ]; then
 
     do_mount_devprocsys
 
+    dmesg_info "Hello from the initramfs"
+
     do_hotplug_scan
 
     # Support /dev/block/mmcXXX only in initrd phase
     ln -s . /dev/block
     ln -s /proc/mounts /etc/mtab
 
+    # Handle whole moto_msm8960_jbbl family automatically
+    DATA_PARTITION="$(fdisk -l /dev/mmcblk0 | awk '/userdata/ { print "/dev/mmcblk0p"$1 }')"
+    dmesg_info "userdata=$DATA_PARTITION"
+
     check_kernel_config
 
     bootsplash
@@ -386,7 +405,9 @@ if [ "$DONE_SWITCH" = "no" ]; then
     echo "" > /proc/sys/kernel/hotplug
 
     if [ -f "/target/init-debug" ]; then
-	exec switch_root /target /init-debug &> /target/init-debug-stderrout
+        adjust_block_device system
+        adjust_block_device cache
+        exec switch_root /target /init-debug &> /target/init-debug-stderrout
     else
 	# Prefer /sbin/preinit over /sbin/init
 	[ -x /target/sbin/preinit ] && INIT=/sbin/preinit || INIT=/sbin/init
@@ -403,6 +424,8 @@ else
 
     do_mount_devprocsys
 
+    dmesg_info "init-debug"
+
     HALT_BOOT="n"
     [ -f /init_enter_debug2 ] && HALT_BOOT="y"
     [ -f /init_disable_telnet ] && DISABLE_TELNET="y"
