From 27525ce53c109583be2e438340780ebad4645ee4 Mon Sep 17 00:00:00 2001
From: elros34 <elros34@users.noreply.github.com>
Date: Sun, 8 Nov 2020 15:13:08 +0000
Subject: [PATCH] Alternative screenshot method

---
 src/screenshotservice.cpp | 18 +++++++-----------
 src/screenshotservice.h   |  3 ++-
 2 files changed, 9 insertions(+), 12 deletions(-)

diff --git a/src/screenshotservice.cpp b/src/screenshotservice.cpp
index f1d95a07..ca969fb7 100644
--- a/src/screenshotservice.cpp
+++ b/src/screenshotservice.cpp
@@ -21,6 +21,8 @@
 #include <QTransform>
 #include <QThreadPool>
 #include <private/qquickwindow_p.h>
+#include <QSharedPointer>
+#include <QQuickItemGrabResult>
 
 #include <unistd.h>
 #include <sys/eventfd.h>
@@ -175,21 +177,15 @@ ScreenshotResult *ScreenshotService::saveScreenshot(const QString &path)
         return result;
     }
 
-    QQuickWindowPrivate *wd = QQuickWindowPrivate::get(compositor);
-    HwcRenderStage *renderStage = (HwcRenderStage *) wd->customRenderStage;
-    if (renderStage)
-        renderStage->setBypassHwc(true);
-
-    QImage grab(compositor->grabWindow());
-
-    if (renderStage) {
-        renderStage->setBypassHwc(false);
-    }
+    auto grabResult = compositor->contentItem()->grabToImage();
 
     const int rotation(QGuiApplication::primaryScreen()->angleBetween(
                 Qt::PrimaryOrientation, compositor->topmostWindowOrientation()));
 
-    QThreadPool::globalInstance()->start(new ScreenshotWriter(notifierId, grab, path, rotation));
+    connect(grabResult.data(), &QQuickItemGrabResult::ready, result, [=]() {
+        QImage grab = grabResult->image();
+        QThreadPool::globalInstance()->start(new ScreenshotWriter(notifierId, grab, path, rotation));
+    });
 
     return result;
 }
diff --git a/src/screenshotservice.h b/src/screenshotservice.h
index e02ea889..3b95a513 100644
--- a/src/screenshotservice.h
+++ b/src/screenshotservice.h
@@ -55,8 +55,9 @@ private:
     Status m_status = Writing;
 };
 
-class ScreenshotService
+class ScreenshotService : public QObject
 {
+    Q_OBJECT
 public:
     static ScreenshotResult *saveScreenshot(const QString &path);
 };
-- 
2.26.2

